app/src/main/java/com/kpiroom/bubble/ui/accountSetup/AccountSetupLogic.kt

```diff
@@ -0,0 +1,182 @@
+    private val isWallpaperSet
+        get() = wallpaperUri.value != defaultWallpaperUri
+
+    val backButtonClicked = MutableLiveData<Boolean>()
+    fun onBackButton() {
+        Source.userPrefs.clear()
+        backButtonClicked.value = true
+    }
+
+    fun onPhotoChanged() {
+        clearUsernameFocus.value = true
+        photoChangeRequested.value = true
+        wallpaperChangeRequested.value = false
+    }
+
+    fun onWallpaperChanged() {
+        clearUsernameFocus.value = true
+        photoChangeRequested.value = false
+        wallpaperChangeRequested.value = true
+    }
+
+    fun onGetStarted() {
+        clearUsernameFocus.value = true
+        progress.apply {
+            username.value.let {
+
+                AsyncProcessor {
+                    progress.loadAsync()
+                    when {
+                        it.isNullOrBlank() -> alertAsync(str(R.string.setup_username_cannot_be_empty))
+
+                        it.length < 6 -> alertAsync(str(R.string.setup_username_too_short))
+
+                        it.length > 16 -> alertAsync(str(R.string.setup_username_too_long))
+
+                        !it.matches(usernameRegex) -> alertAsync(str(R.string.setup_username_invalid))
+
+                        Source.api.usernameExists(it) -> alertAsync(str(R.string.setup_username_exists))
+
+                        !(isPhotoSet && isWallpaperSet) -> progress.alertAsync(
+                            warningPhotosUnchanged,
+                            ::isSetupFinishRequested
+                        )
+
+                        else -> {
+                            finishRequested.postValue(true)
+                            progress.finishAsync()
+                        }
+                    }
+                }
+            } handleError {
+                progress.alert(it.message)
+            } runWith (bag)
+        }
+    }
+
+    private val warningPhotosUnchanged
+        get() = app.getString(
+            R.string.setup_continue_with_unchanged,
+            when {
+                !(isPhotoSet || isWallpaperSet) -> str(R.string.setup_unchanged_photo_and_wallpaper)
+                !isPhotoSet -> str(R.string.setup_unchanged_photo)
+                !isWallpaperSet -> str(R.string.setup_unchanged_wallpaper)
+                else -> null
+            }
+        )
+
+    private fun isSetupFinishRequested(isRequested: Boolean) {
+        if (isRequested)
+            finishRequested.postValue(true)
+        else
+            return
+    }
+
+    fun finishSetup() {
+        progress.apply {
+            AsyncProcessor(Dispatchers.IO) {
+                loadAsync()
+
+                saveUserData()
+                uploadProfile()
+                setupComplete.postValue(true)
+
+                finishAsync()
+            } handleError {
+                alert(it.message)
+            } runWith (bag)
+        }
+    }
+
+    private suspend fun uploadProfile() {
+        Source.userPrefs.let {
```

Comment:
Update this construction without nullable Strings (in UserPrefs and api request)
Make check for `Source.userPrefs.uuid.isNotEmpty()` and if it's empty show alert, cos now user will be confused.